### ISSUE TITLE
[Usage]: relationship between embedding size and vocab_size

### ISSUE THREAD

**Issue by Happy2Git** [vllm-project/vllm#15131]
### Your current environment

```text
The output of `python collect_env.py`
```

### How would you like to use vllm

I’ve noticed that the embedding size is always smaller than the vocab_size. Additionally, sometimes the `prompt_token_ids` are larger than the embedding size. ​Is there a way to map the embedding vector to each of the prompt tokens so that I can retrieve the logit of a prompt token like this:
`embeds[i, labels[i]]`?

```python
outputs = llm.encode(prompts)
print(f'vocab_size: {llm.get_tokenizer().vocab_size}')
for i in range(len(outputs)):
    labels = outputs[i].prompt_token_ids[1:]
    embeds = outputs[i].outputs.data
    print(f'{i}-th prompt_token_ids: {labels}')
    print(f'{i}-th embeddings: {embeds.shape}')
```

```log
Processed prompts: 100%|██████████| 4/4 [00:00<00:00, 55.18it/s, est. speed input: 0.00 toks/s, output: 0.00 toks/s]
vocab_size: 50254
0-th prompt_token_ids: [4007, 273, 253, 1986, 2077, 310]
0-th embeddings: torch.Size([7, 2560])
1-th prompt_token_ids: [13, 619, 1416, 310]
1-th embeddings: torch.Size([5, 2560])
2-th prompt_token_ids: [5347, 273, 6181, 310]
2-th embeddings: torch.Size([5, 2560])
3-th prompt_token_ids: [2852, 273, 14980, 310]
3-th embeddings: torch.Size([5, 2560])
```

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by g-eoj** [comment]
Lets look at:

```
0-th prompt_token_ids: [4007, 273, 253, 1986, 2077, 310]
0-th embeddings: torch.Size([7, 2560])
```

There should be 7 token ids but in the code there is `labels = outputs[i].prompt_token_ids[1:]`. So if we assume no slicing took place, the 7 token ids correspond to embedding vectors. The embeddings object contains 7 vectors of length 2560. So, for example, the embedding for id `2077` is the sixth vector.

This illustration may help, though note it represents tokens as entire words: https://machinelearningmastery.com/wp-content/uploads/2022/01/PE6.png

`vocab_size` is different from the embedding size, and in a certain sense has nothing to do with it. `vocab_size` is a count of all possible token ids.

**Comment by Happy2Git** [comment]
How can I get the logit value of a token with ID `2077` from the embedding vector?

**Comment by g-eoj** [comment]
`embeds[5]`

**Comment by Happy2Git** [comment]
emmm... `embeds[5]` is a vector with length=2560, not a value

**Comment by g-eoj** [comment]
an embedding is a vector... what is it you want to accomplish with it?

**Comment by Happy2Git** [comment]
I wanna use the logits vector to compute cross-entropy loss.

**Comment by g-eoj** [comment]
If I understand right, I think`llm.encode` only returns embeddings, not logits: https://docs.vllm.ai/en/latest/api/offline_inference/llm.html#vllm.LLM.encode

**Comment by Happy2Git** [comment]
You're right. I just wonder how I can use it to calculate logits.

**Comment by g-eoj** [comment]
The only way I know is to hack the `llm.generate` method. I found the logits get discarded here: https://github.com/vllm-project/vllm/blob/9cc645141da1210d88bdf8aa5a7f0c725862b1fd/vllm/model_executor/layers/sampler.py#L278

**Comment by g-eoj** [comment]
Yep! Another idea I haven't tried is to pass a `logits_processor` through the sampling params (https://docs.vllm.ai/en/latest/api/inference_params.html#sampling-parameters). The you wouldn't have to hack vLLM. 

The idea is the logits processor wouldn't modify the logits, but since it has direct access to the logits, you can use it to save the logits as they go through the processor.

**Comment by Happy2Git** [comment]
Thanks! I just tested it. It does work, but it only returns logits for the output tokens - not the prompt ones.
